### PR TITLE
Backport 2.x: Add session ID as an explicit parameter to SSL session cache API

### PR DESCRIPTION
**Pull request by hanno-arm**
This is the backport of the robustness improvement https://github.com/ARMmbed/mbedtls/pull/4347/commits/64ce9741800526fcf62151245f61af10bf269a1e plus later fixes from #4347 to `development_2.x`.

### CODE REVIEW

**Comment by hanno-arm** [comment]
@mpg @yanesca Could you please advise if this backport is enough, or whether there also need to be backports to the LTS branches _at this point_? I am not fully clear yet on the branching strategy.

**Comment by mpg** [comment]
@hanno-arm Thanks for making this backport. We'll want a backport to mbedtls-2.16 too. 

Also, the DCO check points out that one of your commit messages is missing the Signed-Off-By line, can you fix that? Thanks!

File: library/ssl_srv.c [hunk 1]
```diff
@@ -2765,6 +2765,55 @@ static int ssl_write_hello_verify_request( mbedtls_ssl_context *ssl )
 }
 #endif /* MBEDTLS_SSL_DTLS_HELLO_VERIFY */
 
+static void ssl_check_id_based_session_resumption( mbedtls_ssl_context *ssl )
```
Comment (mpg) [inline library/ssl_srv.c:2768]: From the description it looks like you were meaning to push follow-up commits (such as the one that changed this name) - is the current state of the PR what you intended to push?

Comment (hanno-arm) [reply library/ssl_srv.c:2768]: I squashed the follow-up improvements into the commit to keep the history simpler.

Comment (mpg) [reply library/ssl_srv.c:2768]: Ok, I'm fine with squashing. I'm still not happy with the name `ssl_check_xxx` for a function that returns `void`, though :)

Comment (hanno-arm) [reply library/ssl_srv.c]: Fixed.

**Review by mpg — DISMISSED** [review]
LGTM

**Comment by mpg** [comment]
I wonder whether this deserves a ChangeLog entry. In theory it would make a difference if someone was using their own implementation that didn't check, resulting in some session being incorrectly resumed, now those sessions wouldn't be resumed any more. That's an edge case, but it's still observable behaviour.

Note: I don't think that's an incompatible change (in a sense that would prevent from doing it in 2.x or 2.16), in that we're changing from non-compliant to compliant behaviour (in some cases), so it's no more incompatible than any big fix is incompatible.

**Review by yanesca — COMMENTED** [review]
The backport looks good to me too. I think that adding a changelog entry is easier than deliberating wether it is necessary. @hanno-arm would you mind adding a changelog entry?

**Comment by hanno-arm** [comment]
Agree, let's just add it - here it is.

@mpg @yanesca Could you re-review?

**Review by mpg — APPROVED** [review]
Thanks for adding a ChangeLog entry, looks good to me!

**Review by yanesca — APPROVED** [review]
Looks good to me!

**Pull request merged**